python/cudf/cudf/core/_internals/where.py

```diff
@@ -0,0 +1,395 @@
+            raise ValueError(
+                """Replacement list length or number of dataframe columns
+                should be equal to Number of columns of dataframe"""
+            )
+        for i, column_name in enumerate(frame._data.names):
+            input_col = source_df._data[column_name]
+            other_column = others[i]
+            if column_name in cond._data:
+                if isinstance(input_col, cudf.core.column.CategoricalColumn):
+                    if cudf.utils.dtypes.is_scalar(other_column):
+                        try:
+                            other_column = input_col._encode(other_column)
+                        except ValueError:
+                            # When other is not present in categories,
+                            # fill with Null.
+                            other_column = None
+                        other_column = cudf.Scalar(
+                            other_column, dtype=input_col.codes.dtype
+                        )
+                    elif isinstance(
+                        other_column, cudf.core.column.CategoricalColumn
+                    ):
+                        other_column = other_column.codes
+                    input_col = input_col.codes
+
+                result = cudf._lib.copying.copy_if_else(
+                    input_col, other_column, cond._data[column_name]
+                )
+
+                if isinstance(
+                    frame._data[column_name],
+                    cudf.core.column.CategoricalColumn,
+                ):
+                    result = cudf.core.column.build_categorical_column(
+                        categories=frame._data[column_name].categories,
+                        codes=cudf.core.column.as_column(
+                            result.base_data, dtype=result.dtype
+                        ),
+                        mask=result.base_mask,
+                        size=result.size,
+                        offset=result.offset,
+                        ordered=frame._data[column_name].ordered,
+                    )
+            else:
+                out_mask = cudf._lib.null_mask.create_null_mask(
+                    len(input_col),
+                    state=cudf._lib.null_mask.MaskState.ALL_NULL,
+                )
+                result = input_col.set_mask(out_mask)
+            out_df[column_name] = frame[column_name].__class__(result)
+
+        return frame._mimic_inplace(out_df, inplace=inplace)
+
+    else:
+        if isinstance(other, DataFrame):
+            raise NotImplementedError(
+                "cannot align with a higher dimensional Frame"
+            )
+        input_col = frame._data[frame.name]
+        cond = cudf.core.column.as_column(cond)
+        if len(cond) != len(frame):
+            raise ValueError(
+                """Array conditional must be same shape as self"""
+            )
+        if cond.all():
+            result = input_col
```

Comment:
Any reason we needed this instead of just letting it go through the calls and getting the same result? This check is pretty expensive for a generally uncommon case.

Reply:
Removed this check 👍